network/response_parser.c

```diff
@@ -20,80 +22,100 @@ int parse_uri(char *response, http_response *result);
-	int body_index = parse_header(response, result, header_index);
-	result->body = string_from_to(response, body_index, length_pointer_char(response) - 1);
-	return result;
+    char** response_split = string_split(response, "\r\n");
+    int response_split_length = length_pointer_pointer_char(response_split);
+    http_response *result = malloc(sizeof(http_response) + 4 * sizeof(char *));
+    result->header_quantity = count_header(response);
+    if (response_split_length < 2) {
+        free_pointer_pointer_char(response_split);
+        return result;
+    }
+
+    if (result->header_quantity == 0 || !parse_uri(response_split[0], result)) {
+        return result;
+    }
+    register int index;
+    for (index = 0; index < result->header_quantity; index ++) {
+        if (!parse_header(response_split[index + 1], result, index)) {
+            return result;
+        }
+    }
+
+    result->body = strdup("");
+    if (result->header_quantity + 1 < response_split_length) {
+        char* pointer_holder;
+        free(result->body);
+        result->body = strdup(response_split[result->header_quantity + 1]);
+        for (index = result->header_quantity + 2; index < response_split_length; index ++) {
+            pointer_holder = result->body;
+            asprintf(&result->body, "%s\r\n%s", result->body, response_split[index]); // append all data remain to body
+            free(pointer_holder);
+        }
+    }
+    free_pointer_pointer_char(response_split);
+    return result;
 }
 
-int parse_uri(char *response, http_response *result) {
-	register int scan_index = 0;
-	register int mark_index = 0;
-	register int count = 0;
-	while (response[ scan_index ] != '\n' && response[ scan_index ] != '\0') {
-		if (response[ scan_index ] == ' ') {
-			count++;
-			if (count == 1) {
-				result->version = string_from_to(response, mark_index, scan_index - 1);
-				mark_index = scan_index + 1;
-			} else if (count == 2) {
-				result->status_code = string_from_to(response, mark_index, scan_index - 1);
-				mark_index = scan_index + 1;
-			}
-		}
-		scan_index++;
-	}
-	result->status = string_from_to(response, mark_index, scan_index - 1);
-	while (response[ scan_index ] == '\n' || response[ scan_index ] == '\r') {
-		scan_index++;
-	}
-	return scan_index;
+int parse_uri(char *line, http_response *result) {
+    char** uri_split = string_split(line, " ");
+    int uri_split_length = length_pointer_pointer_char(uri_split);
+    if (uri_split_length < 3) {
+        return FALSE;
+    }
+    result->version = strdup(uri_split[0]);
+    result->status_code = strdup(uri_split[1]);
+    result->status = strdup(uri_split[2]);
+    char* pointer_holder = NULL;
+    register int index;
+    for (index = 3; index < uri_split_length; index ++) {
+        pointer_holder = result->status;
+        asprintf(&result->status, "%s %s", result->status, uri_split[index]); // append all text remain to status
+        free(pointer_holder);
+    }
```

Comment:
Some third party response with message has *no body*, we only get result by get `status` or `status code`. 
Version will be use for another purpose. I think get all properties is necessary.
Thanks for review.